syntax = "proto3";

import "ndarray.proto";
import "pingpong.proto";

service LearnerService {
  rpc Persistence(stream Ping) returns (stream Pong);

  rpc GetAction(GetActionRequest) returns (Action);
  rpc GetPolicyVariables(Empty) returns (PolicyVariables);
  rpc GetTDError(GetTDErrorRequest) returns (TDError);
  rpc PostRewards(PostRewardsRequest) returns (Empty);
}

message GetActionRequest {
  NDarray state = 1;
  NDarray rnn_state = 3;
}
message Action {
  NDarray action = 1;
  NDarray rnn_state = 3;
}

message PolicyVariables {
  repeated NDarray variables = 1;
}

message GetTDErrorRequest {
  NDarray n_states = 1;
  NDarray n_actions = 2;
  NDarray n_rewards = 3;
  NDarray state_ = 4;
  NDarray n_dones = 5;
  NDarray n_mu_probs = 6;
  NDarray rnn_state = 7;
}
message TDError {
  NDarray td_error = 1;
}

message PostRewardsRequest {
  NDarray n_rewards = 1;
}